extras/spiffs/spiffs_config.h

```diff
@@ -0,0 +1,263 @@
+#ifndef SPIFFS_GC_HEUR_W_DELET
+#define SPIFFS_GC_HEUR_W_DELET          (5)
+#endif
+// Garbage collecting heuristics - weight used for used pages.
+#ifndef SPIFFS_GC_HEUR_W_USED
+#define SPIFFS_GC_HEUR_W_USED           (-1)
+#endif
+// Garbage collecting heuristics - weight used for time between
+// last erased and erase of this block.
+#ifndef SPIFFS_GC_HEUR_W_ERASE_AGE
+#define SPIFFS_GC_HEUR_W_ERASE_AGE      (50)
+#endif
+
+// Object name maximum length. Note that this length include the
+// zero-termination character, meaning maximum string of characters
+// can at most be SPIFFS_OBJ_NAME_LEN - 1.
+#ifndef SPIFFS_OBJ_NAME_LEN
+#define SPIFFS_OBJ_NAME_LEN             (32)
+#endif
+
+// Size of buffer allocated on stack used when copying data.
+// Lower value generates more read/writes. No meaning having it bigger
+// than logical page size.
+#ifndef SPIFFS_COPY_BUFFER_STACK
+#define SPIFFS_COPY_BUFFER_STACK        (64)
+#endif
+
+// Enable this to have an identifiable spiffs filesystem. This will look for
+// a magic in all sectors to determine if this is a valid spiffs system or
+// not on mount point. If not, SPIFFS_format must be called prior to mounting
+// again.
+#ifndef SPIFFS_USE_MAGIC
+#define SPIFFS_USE_MAGIC                (0)
+#endif
+
+#if SPIFFS_USE_MAGIC
+// Only valid when SPIFFS_USE_MAGIC is enabled. If SPIFFS_USE_MAGIC_LENGTH is
+// enabled, the magic will also be dependent on the length of the filesystem.
+// For example, a filesystem configured and formatted for 4 megabytes will not
+// be accepted for mounting with a configuration defining the filesystem as 2
+// megabytes.
+#ifndef SPIFFS_USE_MAGIC_LENGTH
+#define SPIFFS_USE_MAGIC_LENGTH         (0)
+#endif
+#endif
+
+// SPIFFS_LOCK and SPIFFS_UNLOCK protects spiffs from reentrancy on api level
+// These should be defined on a multithreaded system
+
+// define this to enter a mutex if you're running on a multithreaded system
+#ifndef SPIFFS_LOCK
+#define SPIFFS_LOCK(fs)        // vPortEnterCritical()
+#endif
+// define this to exit a mutex if you're running on a multithreaded system
+#ifndef SPIFFS_UNLOCK
+#define SPIFFS_UNLOCK(fs)      // vPortExitCritical()
+#endif
+
+// Enable if only one spiffs instance with constant configuration will exist
+// on the target. This will reduce calculations, flash and memory accesses.
+// Parts of configuration must be defined below instead of at time of mount.
+#ifndef SPIFFS_SINGLETON
+#define SPIFFS_SINGLETON 1
+#endif
+
+#if SPIFFS_SINGLETON
+// Instead of giving parameters in config struct, singleton build must
+// give parameters in defines below.
+#ifndef SPIFFS_CFG_PHYS_SZ
+#define SPIFFS_CFG_PHYS_SZ(ignore)        (SPIFFS_SIZE)
+#endif
+#ifndef SPIFFS_CFG_PHYS_ERASE_SZ
+#define SPIFFS_CFG_PHYS_ERASE_SZ(ignore)  (4*1024)
+#endif
+#ifndef SPIFFS_CFG_PHYS_ADDR
+#define SPIFFS_CFG_PHYS_ADDR(ignore)      (SPIFFS_BASE_ADDR)
+#endif
+#ifndef SPIFFS_CFG_LOG_PAGE_SZ
+#define SPIFFS_CFG_LOG_PAGE_SZ(ignore)    (256)
```

Comment:
On the top of my head I think a page size of 128 would be better, perhaps
in combination of a logical block size of 2*4*1024. Spiffs will always use
the ```PHYS_ERASE_SZ``` for erasing, so having a 8kB logical block means
spiffs will call erase hal func two times when it needs to erase a block.

Yes, the runtime performance will suffer a bit from having smaller page
sizes, but the flash footprint will be better (i.e. the spiffs structure
will be more filled). I need to verify this to be sure. Perhaps a 256 page
with a 8kB block would be optimal..

2016-07-22 10:33 GMT+02:00 sheinz <notifications@github.com>:

> In extras/spiffs/spiffs_config.h
> <https://github.com/SuperHouse/esp-open-rtos/pull/163#discussion_r71844469>
> :
>
> > +#endif
> > +
> > +#if SPIFFS_SINGLETON
> > +// Instead of giving parameters in config struct, singleton build must
> > +// give parameters in defines below.
> > +#ifndef SPIFFS_CFG_PHYS_SZ
> > +#define SPIFFS_CFG_PHYS_SZ(ignore)        (SPIFFS_SIZE)
> > +#endif
> > +#ifndef SPIFFS_CFG_PHYS_ERASE_SZ
> > +#define SPIFFS_CFG_PHYS_ERASE_SZ(ignore)  (4*1024)
> > +#endif
> > +#ifndef SPIFFS_CFG_PHYS_ADDR
> > +#define SPIFFS_CFG_PHYS_ADDR(ignore)      (SPIFFS_BASE_ADDR)
> > +#endif
> > +#ifndef SPIFFS_CFG_LOG_PAGE_SZ
> > +#define SPIFFS_CFG_LOG_PAGE_SZ(ignore)    (256)
>
> I tried only those values. I thought that logical page size should be the
> same as physical page size for optimal performance.
> The flash chip also supports block erase of 32k and 64k. But it is not
> implemented currently. I'm not sure if we need it.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/SuperHouse/esp-open-rtos/pull/163/files/0ec47b5de9ec705d86cc6d2b210dd791fa6e01a2#r71844469>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE3NV1nuzArxl1_be7cOhOhjg0lxR_s2ks5qYIBHgaJpZM4JNZP1>
> .
>